### PR TITLE
Sockets - Kirsten & Nara

### PR DESCRIPTION
**Pull request by bonara**
# slack.rb
Congratulations! You're submitting your assignment!

You and your partner should _collaborate_ on the answers to these questions.

## Comprehension Questions
Question | Answer
---      | ---
How did you go about exploring the Slack API? Did you learn anything that would be useful for your next project involving an API? | We have read the slack API documentation, tested with postman and googled some of the optional enhancements. We have learned that starting from the documentation is a good idea when working with APIs. 
Give a short summary of the _request/response cycle_. Where does your program fit into that scheme? | Request is when we use HTTParty to get data or send payload through the API. In response to our request, we will get a code indicating whether the request was successful or not and the data if it was a get request. We request data from slack and process it in our program. Users are also allowed to send messages by making a post request to slack. 
How does your program check for and handle errors when using the Slack API? | We check response.code and the ok key in the response. If response ok is false or response.code is not equal to 200, we raise SlackAPIError. 
Did you need to make any changes to the design work we did in class? If so, what were they? | We haven't made changes to the design work we did in class. It worked well for our program. 
Did you use any of the _inheritance idioms_ we've talked about in class? How? | We have used an abstract class and template methods. Our recipient class was an abstract class and details and self.list were the template methods. 
How does VCR aid in testing a program that uses an API? | It reduces the number of API calls you need to make by recording the response. It eliminates the external dependencies for your testing if you have no internet connection or slack is down.

### CODE REVIEW

File: specs/recipient_spec.rb [hunk 1]
```diff
@@ -0,0 +1 @@
+require "test_helper"
```
Comment (kaidamasaki) [inline specs/recipient_spec.rb:1]: This should be `require_relative`, not `require`.  Using `require` here breaks `rake` (but not `guard`).

```suggestion
require_relative "test_helper"
```

File: lib/recipient.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    params = {
+      "token" => ENV["SLACK_API_TOKEN"],
+    }
+    if type == "user"
+      url = USER_URL
+    elsif type == "channel"
+      url = CHANNEL_URL
+    end
+
+    response = HTTParty.get(url, query: params)
```
Comment (kaidamasaki) [inline lib/recipient.rb:29]: Forgot to check for errors.

File: lib/recipient.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    settings = JSON.parse(file)
+    params = {
+      "token" => ENV["SLACK_API_TOKEN"],
+      "channel" => recipient.slack_id,
+      "text" => message,
+      "username" => settings["username"],
+      "icon_emoji" => settings["icon_emoji"],
+    }
+    if recipient.class == User
+      params["as_user"] = true
```
Comment (kaidamasaki) [inline lib/recipient.rb:44]: `as_user` isn't for if the target is a user, it lets you send things as though you are the user whose token you are making the call with.

File: lib/user.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def details
+    return "#{name} (#{real_name}) slack id: #{slack_id}"
+  end
+
+  def self.list
+    raw_data = self.get("user")
+
+    unless raw_data.code == 200
```
Comment (kaidamasaki) [inline lib/user.rb:22]: Forgot to check for the parsed out `ok`.

File: lib/workspace.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    @users = User.list
+    @channels = Channel.list
+    @selected = nil
+  end
+
+  def select_channel(user_input)
+    selected = channels.select do |channel|
+      channel.name == user_input || channel.slack_id == user_input
+    end
+    @selected = selected.first
```
Comment (kaidamasaki) [inline lib/workspace.rb:20]: Ruby's `.find` enumerable method might be helpful to clarify this code.

File: lib/workspace.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def select_user(user_input)
+    selected = users.select do |user|
+      user.name == user_input || user.slack_id == user_input
+    end
+    @selected = selected.first
```
Comment (kaidamasaki) [inline lib/workspace.rb:27]: Same here.

File: lib/recipient.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  def initialize(slack_id, name)
+    @slack_id = slack_id
+    @name = name
+  end
+
+  CHANNEL_URL = "https://slack.com/api/channels.list"
+  USER_URL = "https://slack.com/api/users.list"
+  POST_URL = "https://slack.com/api/chat.postMessage"
+  def self.get(type)
```
Comment (kaidamasaki) [inline lib/recipient.rb:19]: `self.get` should take `url` and `params` not just a `type`.  Keeping track of what to pass in should be handled by `Workspace`.

File: lib/recipient.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  attr_accessor :slack_id, :name
+
+  def initialize(slack_id, name)
+    @slack_id = slack_id
+    @name = name
+  end
+
+  CHANNEL_URL = "https://slack.com/api/channels.list"
+  USER_URL = "https://slack.com/api/users.list"
+  POST_URL = "https://slack.com/api/chat.postMessage"
```
Comment (kaidamasaki) [inline lib/recipient.rb:18]: URL constants!  🎉 

Though, I think these probably wanted to live in `Workspace`.

File: specs/recipient_spec.rb (hunk 1 above)
Comment (kaidamasaki) [reply specs/recipient_spec.rb:1]: Also, while it's hard to test an abstract class it would be good to at a minimum make sure that your template methods `raise` `NotImplementedError`.

File: specs/test_helper.rb [hunk 9]
```diff
@@ -1,15 +1,31 @@
...
\ No newline at end of file
+  config.cassette_library_dir = "specs/cassettes" # folder where casettes will be located
+  config.hook_into :webmock # tie into this other tool called webmock
+  config.default_cassette_options = {
+    :record => :new_episodes,    # record new data when we don't have it yet
+    :match_requests_on => [:method, :uri, :body], # The http method, URI and body of a request all need to match
+  }
+  # Don't leave our token lying around in a cassette file.
+  config.filter_sensitive_data("<SLACK_API_TOKEN>") do
+    ENV["SLACK_API_TOKEN"]
```
Comment (kaidamasaki) [inline specs/test_helper.rb:29]: 🎉

File: specs/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        expect(response["members"].first["id"]).must_equal "USLACKBOT"
+      end
+    end
+
+    # it "raises an exception if response is invalid" do
+    #   VCR.use_cassette("slack_user") do
+    #     response = User.get("user")
+    #     expect(response["ok"]).must_equal false
+    #   end
+    # end
```
Comment (kaidamasaki) [inline specs/user_spec.rb:20]: My guess is you couldn't get this functionality working in time for submission.  

It was a good test though and would have caught an issue.  👍🏼

File: specs/channel_spec.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+require_relative "test_helper"
+
+describe "Channel class" do
```
Comment (kaidamasaki) [inline specs/channel_spec.rb:3]: I would have liked to see some edge case tests on `Channel`.  What happens if you try to get a channel that doesn't exist, etc?

File: lib/workspace.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def show_details
+    return @selected.details
+  end
+
+  def print_details(recipients)
+    if recipients == "users"
```
Comment (kaidamasaki) [inline lib/workspace.rb:35]: I'd like to see `recipients` just be an array of objects, but if you want it to be something you switch on like this you should use `:users` instead of `"users"` since generally we want to use atoms when we match on things like this.

File: lib/workspace.rb [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    elsif recipients == "channels"
+      return_array = []
+      channels.each do |channel|
+        return_array << channel.details
+      end
+    end
+    return return_array
+  end
+
+  def update_settings(name, emoji)
```
Comment (kaidamasaki) [inline lib/workspace.rb:49]: I like that your configuration system is persistent and uses JSON. 😃

File: specs/workspace_spec.rb [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    it "returns text from print_details" do
+      VCR.use_cassette("slack_workspace") do
+        workspace = Workspace.new
+        expect(workspace.print_details("users")).must_be_kind_of Array
+
+        expect(workspace.print_details("channels")).must_be_kind_of Array
+      end
+    end
+
+    it "returns a user object from select_user" do
```
Comment (kaidamasaki) [inline specs/workspace_spec.rb:25]: What if the user you try to select doesn't exist? What happens to a previously selected user if you try to select a bad user?

File: specs/workspace_spec.rb [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+      VCR.use_cassette("slack_workspace") do
+        workspace = Workspace.new
+        expect(workspace.select_user("slackbot")).must_be_kind_of User
+
+        expect(workspace.select_user("USLACKBOT")).must_be_kind_of User
+      end
+    end
+
+    it "returns a channel object from select_channel" do
```
Comment (kaidamasaki) [inline specs/workspace_spec.rb:34]: Same for `select_channel`.

**Comment by kaidamasaki** [comment]
# slack.rb
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene (no slack tokens) | 👍🏼
Comprehension questions | 👍🏼
**Functionality** |
List users/channels | 👍🏼
Select user/channel | 👍🏼
Show details | 👍🏼
Send message | 👍🏼
Program runs without crashing | 👍🏼
**Implementation** |
API errors are handled appropriately | In some places; in others it looks like you forgot to check.
Inheritance model matches in-class activity | Mostly.
Inheritance idioms (abstract class, template methods, polymorphism) used appropriately | 👍🏼
Methods are used to break work down into simpler tasks | 👍🏼 `slack.rb` could have been cleaned up a little but overall it looks good!
Class and instance methods are used appropriately | 👍🏼
**Overall** | Great job!  There are a few rough edges (I feel like maybe you were pressed for time) but overall it's a really solid solution with good, clean code.
